common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/SmartcardNfcCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,175 @@
+                                tryUsingSmartcardWithPin(pin, certDetails, request, session);
+                                clearPin(pin);
+                            }
+
+                            @Override
+                            public void onException(@NonNull final Exception e) {
+                                indicateGeneralException(methodTag, e);
+                                request.cancel();
+                                clearPin(pin);
+                                mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+                            }
+                        });
+                    }
+
+                    @Override
+                    public void onClosedConnection() {
+                        //Nothing needed.
+                    }
+                });
+                mSmartcardCertBasedAuthManager.startDiscovery(mActivity);
+            }
+        };
+    }
+
+    /**
+     * Checks to see if PIN for smartcard is correct.
+     * If so, proceed to attempt authentication.
+     * Otherwise, handle the incorrect PIN based on how many PIN attempts are remaining.
+     * @param pin char array containing PIN attempt.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @param session An ISmartcardSession created to help with interactions pertaining to certificates.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    @Override
+    protected void tryUsingSmartcardWithPin(@NonNull final char[] pin,
+                                            @NonNull final ICertDetails certDetails,
+                                            @NonNull final ClientCertRequest request,
+                                            @NonNull final ISmartcardSession session)
+            throws Exception {
+        final String methodTag = TAG + ":tryUsingSmartcardWithPin";
+        if (session.verifyPin(pin)) {
+            //If pin is successfully verified, we will use the certificate to perform the rest of the logic for authentication.
+            useSmartcardCertForAuth(certDetails, pin, session, request);
+            return;
+        }
+        final int attemptsRemaining = session.getPinAttemptsRemaining();
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+        // If the number of attempts is 0, no more attempts will be allowed.
+        if (attemptsRemaining == 0) {
+            //We must display a dialog informing the user that they have made too many incorrect attempts,
+            // and the user will need to figure out a way to reset their key outside of our library.
+            indicateTooManyFailedAttempts(methodTag);
+            request.cancel();
+            return;
+        }
+        mDialogHolder.showPinDialog(
```

Comment:
do we need this? (since setPinDialogErrorMode will be invoked right below)

Reply:
Yes, since unlike the usb implementation, the nfc flow includes an extra two dialogs after the pin dialog (one for prompting user to tap and another telling them to hold their smartcard) , so the displayed dialog is no longer the pin dialog. The pin dialog needs to be displayed again to the user, and then setPinDialogErrorMode can be called (note that setPinDialogErrorMode does not create and display the pin dialog, it just sets some ui properties).